dask/array/linalg.py

```diff
@@ -328,3 +328,117 @@ def svd(a, name=None):
+
+def _lu_vertical(arr, u):
+    return np.dot(arr, np.linalg.inv(u))
+
+def _dummy_identity(x):
+    # ToDo: Remove
+    return x
+
+def lu(a):
+    import scipy.linalg
+
+    if a.ndim != 2:
+        raise ValueError
+
+    vdim = len(a.chunks[0])
+    hdim = len(a.chunks[1])
+
+    token = tokenize(a)
+    name_lu = 'lu-lu-' + token
+
+    # ToDo: Remove
+    name_l_internal = 'lu-l-internal-' + token
+    name_u_internal = 'lu-u-internal-' + token
+
+    name_p = 'lu-p-' + token
+    name_l = 'lu-l-' + token
+    name_u = 'lu-u-' + token
+
+    dsk = {}
+
+    for i in range(min(vdim, hdim)):
+        if i == 0:
+            dsk[(name_lu, i, i)] = (scipy.linalg.lu, (a.name, i, i))
+            for j in range(i + 1, hdim):
+                # sweep to horizontal
+                dsk[(name_lu, i, j)] = (_lu_horizontal, (a.name, i, j), (name_l_internal, i, i))
+            for k in range(i + 1, vdim):
+                # sweep to vertical
+                dsk[(name_lu, k, i)] = (_lu_vertical, (a.name, k, i), (name_u_internal, i, i))
+        else:
+            dsk[(name_lu, i, i)] = (_lu_diag, (a.name, i, i), (name_l_internal, i, i - 1),
+                                    (name_u_internal, i - 1, i))
+
+            for j in range(i + 1, hdim):
+                # sweep to horizontal
+                dsk[(name_lu, i, j)] = (_lu_horizontal_diag, (a.name, i, j),
+                                                             (name_l_internal, i, 1),
+                                                             (name_u_internal, i - 1, j),
+                                                             (name_l_internal, i, i))
+            for k in range(i + 1, vdim):
+                # sweep to vertical
+                dsk[(name_lu, k, i)] = (_lu_vertical_diag, (a.name, k, i),
+                                                           (name_l_internal, k, 1),
+                                                           (name_u_internal, 1, i)
+                                                           (name_u_internal, i, i))
+
+        for j in range(i + 1, hdim):
+            # sweep to horizontal
+            dsk[(name_lu, i, j)] = (_lu_horizontal, (a.name, i, j), (name_l_internal, i, i))
+        for k in range(i + 1, vdim):
+            # sweep to vertical
+            dsk[(name_lu, k, i)] = (_lu_vertical, (a.name, k, i), (name_u_internal, i, i))
+
+    for i in range(min(vdim, hdim)):
+        for j in range(min(vdim, hdim)):
+            if i == j:
+                dsk[(name_l_internal, i, j)] = (operator.getitem, (name_lu, i, j), 1)
+                dsk[(name_u_internal, i, j)] = (operator.getitem, (name_lu, i, j), 2)
+
+                dsk[(name_p, i, j)] = (operator.getitem, (name_lu, i, j), 0)
+
+            elif i > j:
+                # ToDo: remove _dummy_identity
+                dsk[(name_p, i, j)] = (np.zeros, (a.chunks[0][i], a.chunks[1][j]))
+                dsk[(name_l_internal, i, j)] = (_dummy_identity, (name_lu, i, j))
```

Comment:
@mrocklin Yes your example should work, and #886 intends to the case like below. Because calculated ``L`` and ``U`` is referred by intermediate calculation.

```
dsk = {('x', 0): np.array([1, 2]),
       ('x', 1): ('x', 0)}
darr = da.Array(dsk, 'x', chunks=(2, ), shape=(4, ))
darr.compute()
# KeyError: ('x', 0)
```


@shoyer Thanks, will fix that.